### DISCUSSION TITLE
useTranslation Hook not showing autocomplete for Next.js (TS)

### DISCUSSION THREAD

**Discussion by abdulrahimiliasu** [i18next/next-i18next#2029] [Q&A] [answered]
When importing `useTranslation` hook from `next-18next` module, it does not provide autocompletion (types), however when imported from `react-i18next` it does show autocomplete suggestions. So why don't I use `react-i18next` instead right ? well the problem is `useTranslation` hook imported from `react-i18next` does not show the content on the page, however  `next-18next` does.

here is my `i18next.d.ts` living in my **types** folder

```typescript
/**
 * If you want to enable locale keys typechecking and enhance IDE experience.
 *
 * Requires `resolveJsonModule:true` in your tsconfig.json.
 *
 * @link https://www.i18next.com/overview/typescript
 */
import "i18next";

import type common from "../../public/locales/en/common.json";
import type main from "../../public/locales/en/main.json";
import type index from "../../public/locales/en/index.json";
import type login from "../../public/locales/en/login.json";
import type signup from "../../public/locales/en/signup.json";

interface I18nNamespaces {
    common: typeof common;
    main: typeof main;
    index: typeof index;
    login: typeof login;
    signup: typeof signup;
}

declare module "i18next" {
    interface CustomTypeOptions {
        defaultNS: "common";
        resources: I18nNamespaces;
    }
}
```

and here is my `index.ts` page for example

```typescript
import { useTranslation } from "next-i18next"; // works but no autocomplete
// import { useTranslation } from "react-i18next"; // does not show content instead ( e.g common:login) but  autocomplete exist


type PageProps = {};

export default function Home({}: InferGetServerSidePropsType<GetServerSideProps>) {
    const { t } = useTranslation(["common", "index"]);

    return (
        <div>
            <TitleText>{strings.app_name}</TitleText>
            <ThemeSwitcher variant="multiple" />

            <Link href={"/login"}>
                <u>{t("common:login")}</u>
            </Link>
            <br />
            <Link href={"/signup"}>
                <u>{t("common:create-an-account")}</u>
            </Link>
        </div>
    );
}

export const getServerSideProps: GetServerSideProps<PageProps> = async ({ locale }) => {
    return {
        props: {
            ...(await serverSideTranslations(locale ?? "en", ["common", "index"])),
        },
    };
};
```

Please any help or suggestion will be appreciated

**Comment by belgattitude** [comment, marked as answer]
Upcoming release only (v13.0.0). In the meantime 

```typescript
import "react-i18next"; // (in the future 'i18next')

import type common from "../../public/locales/en/common.json";
import type main from "../../public/locales/en/main.json";
import type index from "../../public/locales/en/index.json";
import type login from "../../public/locales/en/login.json";
import type signup from "../../public/locales/en/signup.json";

interface I18nNamespaces {
    common: typeof common;
    main: typeof main;
    index: typeof index;
    login: typeof login;
    signup: typeof signup;
}

// (was 'i18next')
declare module "react-i18next" {
    interface CustomTypeOptions {
        defaultNS: "common";
        resources: I18nNamespaces;
    }
}
```

When 13.0.0 is out, you'll have to change once more :) See [UPGRADING](https://github.com/i18next/next-i18next/blob/master/UPGRADING.md).

Let me know if it helped

**Reply by abdulrahimiliasu** [reply]
Still does not show auto complete on either imported hooks

**Reply by belgattitude** [reply]
@abdulrahimiliasu difficult to know without the source. but I can confirm both v12 and v13 ([available now](https://github.com/i18next/next-i18next/releases/tag/v13.0.0)) works in my setups (ie: https://github.com/belgattitude/nextjs-monorepo-example/pull/2932).

Could you confirm

1. [tsconfig] `resolveJsonModule: true`
2. [tsconfig] ensure json is well included, for example

```typescript
"include": [
    "next-env.d.ts",
    "**/*.ts",
    "**/*.tsx",
    "**/*.mts",
    "**/*.js",
    "**/*.cjs",
    "**/*.mjs",
    "**/*.jsx",
    "**/*.json"
  ]
```

3. Ensure the name of the dts file is well `./src/@types/react-i18next.d.ts` (on < 13) or `./src/@types/i18next.d.ts` (>= 13). (.src)
4. Ensure that `"../../public/locales/en/signup.json";` path can be properly resolved to a valid  path from `tsconfig.compilerOptions.baseUrl` (typescript won't complain if it can't resolve a json file and will put any in place).

**Reply by abdulrahimiliasu** [reply]
Here is my `tsconfig` file, as you can see everything there seems okay,

```typescript
{
    "compilerOptions": {
        "target": "es5",
        "lib": ["dom", "dom.iterable", "esnext"],
        "allowJs": true,
        "skipLibCheck": true,
        "strict": true,
        "forceConsistentCasingInFileNames": true,
        "noEmit": true,
        "esModuleInterop": true,
        "module": "esnext",
        "moduleResolution": "node",
        "resolveJsonModule": true,
        "isolatedModules": true,
        "jsx": "preserve",
        "incremental": true,
        "rootDir": ".",
        "baseUrl": "./src",
        "paths": {
            "@abstractions/*": ["abstractions/*"],
            "@components/*": ["components/*"],
            "@constants/*": ["constants/*"],
            "@context/*": ["context/*"],
            "@hooks/*": ["hooks/*"],
            "@layout/*": ["layout/*"],
            "@repository/*": ["repository/*"],
            "@theme/*": ["theme/*"],
            "@utils/*": ["utils/*"],
            "@public/*": ["./public/*"],
            "types/*": ["types/*"]
        },
        "types": ["node"]
    },
    "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", "**/*.mts", "**/*.js", "**/*.cjs", "**/*.mjs", "**/*.jsx", "**/*.json"],
    "exclude": ["node_modules"]
}
```

the `react-18next.d.ts` file is in `<root project dir>/src/types/react-18next.d.ts`
the public folder path `<root project dir>/public/locales/en/...files`

**Reply by abdulrahimiliasu** [reply]
Auto complete will work if I change the module augmentation from `"react-i18next"` to `"i18next"` and the use `useTranslation()` from `react-i18next` and not `next-i18next`. But then the content will not show up in the site :(

**Reply by abdulrahimiliasu** [reply]
But as of now, I only see `"translation"` from auto complete when using `useTranslation()` hook from `next-i18next`, and also can't see any auto complete when using `t("")`

**Reply by belgattitude** [reply]
Ok that's something. Tried to test on my example repo and can't reproduce. 

Both

```
import { useTranslation } from 'next-i18next';
import { useTranslation } from 'react-i18next';
```

Works (on v13.0.0).

![image](https://user-images.githubusercontent.com/259798/202857077-22f5753d-eb47-4874-9ebf-5fe7efc9081b.png)

(from [nextjs-monorepo-example](https://github.com/belgattitude/nextjs-monorepo-example/blob/main/apps/nextjs-app/src/features/home/pages/HomePage.tsx))

That said, I feel if you were on v12 there's another hypothesis. Your package manager has updated or has duplicates of react-i18next>22 and i18next>12 (you can check in the lock file).  See also: https://github.com/i18next/next-i18next/blob/master/TROUBLESHOOT.md#how-to-debug-installation

Note that the possibility of duplicates is better handled in v13.0.0. 

Could you try an upgrade ? https://github.com/i18next/next-i18next/releases/tag/v13.0.0

**Reply by abdulrahimiliasu** [reply]
Oh I forgot to mention, I'm currently on v12. my `Package.json file`
```json
"dependecies" : 
{
       "i18next": "^22.0.6",
        "next": "13.0.0",
        "next-i18next": "^12.1.0",
        "react": "18.2.0",
        "react-dom": "18.2.0",
        "react-i18next": "^12.0.0"
}
```

**Reply by belgattitude** [reply]
Then that makes sense. You'll probably have i18next installed twice (21.x and 22.x), same for react-i18next. And that will create issues.

Let me know if it still happen when upgrading `"next-i18next": "^13.0.0"`. Hope it will solve it

**Reply by abdulrahimiliasu** [reply]
Wow, that literally solved everything 😄 , Thanks a lot 👍

**Reply by belgattitude** [reply]
Top :smile:
